vignettes/qenv.Rmd

```diff
@@ -0,0 +1,144 @@
+```
+### Joining `qenv` objects
+
+Given a pair of `qenv` objects it might be possible to "join" them together creating a new `qenv` object containing the union of both environments together with the code needed to reproduce it:
+
+```{r}
+
+common_q <- eval_code(new_qenv(), quote(x <- 1))
+
+x_q <- eval_code(common_q, quote(y <- 5))
+y_q <- eval_code(common_q, quote(z <- 5))
+
+join_q <- join(x_q, y_q)
+
+print(join_q)
+```
+
+Dependent on the contents of the environments and order of code it may not be possible to join `qenv` objects. See the function documentation for more details.
+
+### Warnings and messages in `qenv` objects
+
+If warnings or messages are thrown when evaluating code in a `qenv` environment, they are captured and stored in the  `qenv` object. These messages and warnings can be accessed with the `@` operator.
+
+```{r}
+q_message <- eval_code(new_qenv(), quote(message("this is a message")))
+q_message@messages
+
+q_warning <- eval_code(new_qenv(), quote(warning("and this is a warning")))
+q_warning@warnings
+```
+
+If no warning or message occurs on a particular line of code, the corresponding message/warning value will be an empty string.
+
+```{r}
+q_message@warnings
+q_warning@messages
+```
+
+A helper function `get_warnings()` is also available to produce a formatted string containing the warnings and the code used to produce them - it returns NULL when there are no warnings.
+
+## Using `qenv` inside `shiny` applications
+
+The functions above can be combined into a `shiny` application generating reproducible outputs. In the example below the `verbatim_popup` from `teal.widgets` is used to hide the code used to generate output in a modal dialog. 
+When using a `qenv` to evaluate code, if an error occurs then an object of type `qenv.error` is created. This object can be used in all the same places a `qenv` object is used and means you do not need to change your code to handle these errors - select `error_option` in the example below to see `qenv` error handling in action. 
+
+```{r}
+library(shiny)
+library(teal.widgets)
+
+# create an initial qenv with the data in
+data_q <- new_qenv() %>% eval_code("iris_data <- iris")
+
+ui <- fluidPage(
+  radioButtons(
+    "option", "Choose a column to plot:",
+    c("Sepal.Length", "Sepal.Width", "Petal.Length", "Petal.Width", "error_option")
+  ),
+  # button for the Code to be shown when clicked on
+  verbatim_popup_ui("rcode", button_label = "Show Code"),
+  plotOutput("plot")
+)
+
+server <- function(input, output, session) {
+  # create a qenv containing the reproducible output
+  output_q <- reactive({
+    req(input$option)
+    eval_code(
+      data_q,
+      bquote(p <- hist(iris_data[, .(input$option)]))
+    )
+  })
+
+  # display output
+  output$plot <- renderPlot(output_q()[["p"]])
+
+  # code is displayed when "Show Code" button is clicked on
+  verbatim_popup_srv("rcode", reactive(paste(get_code(output_q()), collapse = "\n")), title = "Code")
```

Comment:
@gogonzo do we actually need the paste and collapse here?
```suggestion
  verbatim_popup_srv("rcode", reactive(get_code(output_q())), title = "Code")
```

Reply:
I think it's needed for renderText -> verbatimTextOutput. I'll check it

Reply:
Seems to work without it -> verbatim popup must be smart enough